lib/rust/lazy-reader/src/decoder.rs

```diff
@@ -0,0 +1,168 @@
+#![allow(unsafe_code)]
+
+//! This module exports various UTF decoders for decoding UTF32 characters.
+
+use std::fmt::Debug;
+
+
+
+// ===============
+// === Decoder ===
+// ===============
+
+/// The result of `decoder.decode`.
+#[derive(Debug,Clone,Copy)]
+pub struct Char {
+    /// The decoded character.
+    pub char: Option<char>,
+    /// The number of words read.
+    pub size: usize,
+}
+
+/// Trait for decoding UTF32 characters.
+pub trait Decoder {
+    /// The input of the decoder.
+    type Word : Default + Copy + Debug;
+    /// The maximum amount of words needed to decode one symbol.
+    const MAX_SYMBOL_LEN: usize;
```

Comment:
Can we call this `MAX_CODEPOINT_LEN` instead? We're decoding codepoints with this specifically, so it would be great to be precise.